lib/appium.js

```diff
@@ -344,35 +346,142 @@ class AppiumDriver extends BaseDriver {
+    // by some but not all plugins, or by plugin(s) but not by the default behavior. So start out
+    // this object declaring that the default handler has not been executed.
+    const handled = {default: false};
+
+    // now we define an async function which will be passed to plugins, and successively wrapped
+    // if there is more than one plugin that can handle the command. To start off with, the async
+    // function is defined as calling the default behavior, i.e., whichever of the 3 cases above is
+    // the appropriate one
+    const defaultBehavior = async () => {
+      log.info(`Executing default handling behavior for command '${cmd}'`);
+      handled.default = true; // if we make it here, we know that the default behavior is handled
+
+      if (isGetStatus) {
+        return await this.getStatus();
+      }
 
-    const sessionId = _.last(args);
-    const dstSession = await sessionsListGuard.acquire(AppiumDriver.name, () => this.sessions[sessionId]);
-    if (!dstSession) {
-      throw new Error(`The session with id '${sessionId}' does not exist`);
-    }
+      if (isUmbrellaCmd) {
+        // some commands, like deleteSession, we want to make sure to handle on *this* driver,
+        // not the platform driver
+        return await super.executeCommand(cmd, ...args);
+      }
 
-    let res = {
-      protocol: dstSession.protocol
+      // here we know that we are executing a session command, and have a valid session driver
+      return await dstSession.executeCommand(cmd, ...args);
     };
 
+    // now take our default behavior, wrap it with any number of plugin behaviors, and run it
+    const wrappedCmd = this.wrapCommandWithPlugins({cmd, args, handled, next: defaultBehavior});
+    const res = await this.executeWrappedCommand({wrappedCmd, protocol});
+
+    this.logPluginHandlerReport({cmd, handled});
+
+    return res;
+  }
+
+  wrapCommandWithPlugins ({cmd, args, next, handled}) {
+    // get any plugins which are registered as handling this command
+    const plugins = this.pluginsToHandleCmd(cmd);
+
+    plugins.length && log.info(`Plugins which can handle cmd '${cmd}': ${plugins.map((p) => p.name)}`);
+
+    // now we can go through each plugin and wrap `next` around its own handler, passing the *old*
+    // next in so that it can call it if it wants to
+    for (const plugin of plugins) {
+      // need an IIFE here because we want the value of next that's passed to plugin.handle to be
+      // exactly the value of next here before reassignment; we don't want it to be lazily
+      // evaluated, otherwise we end up with infinite recursion of the last `next` to be defined.
+      handled[plugin.name] = false; // we see a new plugin, so add it to the 'handled' object
+      next = ((_next) => async () => {
+        log.info(`Plugin ${plugin.name} is now handling cmd '${cmd}'`);
+        handled[plugin.name] = true; // if we make it here, this plugin has attempted to handle cmd
```

Comment:
how do we guarantee that plugin names are unique? I think it would be more logical to generate uuid for each plugin and use it as a key

Reply:
Plugin names are guaranteed to be unique because the`appium plugin install` command won't let you install one if another one with the same name is installed. Plugin names are also keys in the yaml, so even if you edit it by hand no duplicates should show up. It's up to the ecosystem to not step on each other's toes (or as a user, you need to choose which plugin of two that register identical names you want to use).